Disable CocoaPods input and output paths in Xcode build phase and adopt new Xcode build system

jmagman opened the pull request:
## Description

Previously the Xcode new build system could not be adopted due to two Xcode project build phases both outputting the Flutter.framework (one for Flutter w/o packages and the other from CocoaPods).  The legacy build system was more lenient about duplicate outputs.  See https://github.com/flutter/flutter/issues/20685#issuecomment-421511890 for a good description of the problem.

This change updates the Podfile templates to use the CocoaPod disable_input_output_paths installation option, which prevents the [CP] Embed Pods Frameworks build phase from outputting the Flutter.framework files.  This has the unfortunate side-effect of also preventing any input files, which means the Pods-Runner-frameworks.sh script will run on every incremental build.

This change also removes the workaround from https://github.com/flutter/flutter/pull/21901/commits/231784ecf6a38b47ec1994609da1bbbb3fe5e99a opting into the Xcode legacy build system.

## Related Issues

Fixes https://github.com/flutter/flutter/issues/20685.

## Tests

1. Updated the CocoaPods version tests to check for the new minimum required version where disable_input_output_paths was introduced.
2. Added a check to ensure the WorkspaceSettings file is no longer generated.

I manually tested the `flutter create` flow and confirmed the generated Xcode project builds. Then I added packages and confirmed that Xcode project builds.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [x] No, this is *not* a breaking change.

jmagman committed:
Disable Cocoapods input and output paths in Xcode build phase and adopt new Xcode build system

jonahwilliams commented:
I noticed that this also required changes to some of the existing generated config files, how will this affect apps which have not been updated?

- Do we also need to update the templates so that new applications are set up correctly?

jmagman commented:
> I noticed that this also required changes to some of the existing generated config files, how will this affect apps which have not been updated?
Existing apps that are currently building using the legacy build system will continue to use the legacy build system.  No updates will be needed from developers.  This will only impact newly generated apps.  I updated the examples so no one would crack open those Xcode projects and notice the legacy build system and get confused.

I am reluctant to migrate the build system under an existing app since that build system is stricter, and we could be exposing new compilation issues.  It's definitely more risky, but if you think it's worth it @jonahwilliams and @cbracken, I can try.  It might be annoying to developers that they updated flutter, then have to track down new Xcode compilation errors when their .dart code didn't change (though I suppose that probably exact scenario happens nearly every time Xcode is updated).

> * Do we also need to update the templates so that new applications are set up correctly?
I updated the Podfile templates so new applications are set up correctly.  I tested by `flutter create app` then `flutter build ios` and made sure the Xcode workspace builds.  Then I added a package, rebuilt, and confirmed the Xcode workspace builds, and that the [CP] Embed Pods Frameworks build phase does not have input or output files.

Is there another template I'm missing?

jonahwilliams commented:
> Is there another template I'm missing?

Ahh, no that's fine  - I just missed it

jonahwilliams commented:
> . This has the unfortunate side-effect of also preventing any input files, which means the Pods-Runner-frameworks.sh script will run on every incremental build.

Is there any path forward here?

jmagman commented:
> > . This has the unfortunate side-effect of also preventing any input files, which means the Pods-Runner-frameworks.sh script will run on every incremental build.
> 
> Is there any path forward here?

Not really... @cbracken and I discussed this and we think it's an okay medium-term solution, and a trade-off that's worth getting off the legacy build system in case Xcode pulls the plug.

From https://github.com/flutter/flutter/issues/20685#issuecomment-497501567
>If we get to the point where we're seeing plugin Cocoapods that are really expensive to compile, we can deal with that as an optimisation.

jmagman committed:
deploy_gallery-macos test should install CocoaPods

stuartmorgan-g reviewed (approved):
LGTM. I confirmed that the macOS plugin flow works with the added line (and that it eliminates the strident warning in the Xcode UI about duplicate output files for the current macOS runner).

jmagman commented:
Updates the Podfile template to use the CocoaPod disable_input_output_paths installation option which prevents the [CP] Embed Pods Frameworks build phase from outputting the Flutter.framework files.

jmagman commented:
This change increased the minimum required version of CocoaPods, which caused a build failure on macOS slaves running an older version.  I reverted this change with https://github.com/flutter/flutter/commit/9734f4ed43eef148a354ac4f1e7118883434e2d3 while the new version requirement configuration propagates through the CI system.
I will un-revert tomorrow.

jmagman commented:
Un reverted with https://github.com/flutter/flutter/commit/d053fe5cb3967288363109d426bb1e0692ecf0cd.
